perceval/backends/github.py

```diff
@@ -0,0 +1,250 @@
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+#
+# Authors:
+#   Alvaro del Castillo San Felix <acs@bitergia.com>
+#
+
+
+
+import json
+import logging
+import os.path
+
+import requests
+
+from ..backend import Backend, BackendCommand
+from ..cache import Cache
+from ..errors import CacheError
+from ..utils import DEFAULT_DATETIME, str_to_datetime
+
+logger = logging.getLogger(__name__)
+
+class GitHub(Backend):
+    """GitHub backend for Perceval
+
+    This class allows the fetch the issues stored in GitHub
+    repository.
+    """
+
+    def __init__(self, owner=None, repository=None, token=None,
+                 cache=None):
+        """ Init a GitHub instance
+        :param owner: GitHub owener
+        :param repository: GitHub repository from the owner
+        :param token: GitHub auth token to access the API
+        :param cache: Use issues already retrieved in cache
+        """
+        super().__init__(cache=cache)
+        self.owner = owner
+        self.repository = repository
+        self.client = GitHubClient(owner, repository, token)
+
+    def fetch(self, from_date=DEFAULT_DATETIME):
+        """Fetch the issues from the repository.
+
+        The method retrieves, from a GitHub repository, the issues
+        updated since the given date.
+
+        :param from_date: obtain issues updated since this date
+
+        :returns: a generator of issues
+        """
+
+        self._purge_cache_queue()
+
+        issues = self.client.get_issues(from_date)
+
+        self._push_cache_queue(issues)
+        self._flush_cache_queue()
+
+        while issues:
+            issue = issues.pop(0)
+            yield issue
+
+            if not issues:
+                issues = self.client.get_issues()
+                if issues and len(issues)>0:
+                    self._push_cache_queue(issues)
+                    self._flush_cache_queue()
+
+    def fetch_from_cache(self):
+        """Fetch the issues from the cache.
+
+        It returns the issues stored in the cache object provided during
+        the initialization of the object. If this method is called but
+        no cache object was provided, the method will raise a `CacheError`
+        exception.
+
+        :returns: a generator of items
+
+        :raises CacheError: raised when an error occurs accessing the
+            cache
+        """
+        if not self.cache:
+            raise CacheError(cause="cache instance was not provided")
+
+        cache_items = self.cache.retrieve()
+
+        for issues in cache_items:
+            for issue in issues:
+                yield issue
+
+
+class GitHubClient:
+
+    def __init__(self, owner, repository, token):
+        self.owner = owner
+        self.repository = repository
+        self.auth_token = token
+        self.last_page = self.page = 1  # pagination in items downloading
```

Comment:
Don't get used to declare variables this way. Here there's no problem because `1` is an immutable object but it won't work the same when the objects are mutable. For instance:
```
>>> a = b = []
>>> a.append('x')
>>> print(a)
['x']
>>> print(b)
['x']
```